nav2_util/include/nav2_util/service_client.hpp

```diff
@@ -12,22 +12,29 @@
 // See the License for the specific language governing permissions and
 // limitations under the License. Reserved.
 
-#ifndef NAV2_TASKS__SERVICE_CLIENT_HPP_
-#define NAV2_TASKS__SERVICE_CLIENT_HPP_
+#ifndef NAV2_UTIL__SERVICE_CLIENT_HPP_
+#define NAV2_UTIL__SERVICE_CLIENT_HPP_
 
 #include <string>
 #include "rclcpp/rclcpp.hpp"
+#include "nav2_util/node_utils.hpp"
 
-namespace nav2_tasks
+namespace nav2_util
 {
 
 template<class ServiceT>
 class ServiceClient
 {
 public:
-  explicit ServiceClient(const std::string & name)
+  explicit ServiceClient(
+    const std::string & name,
+    const rclcpp::Node::SharedPtr & provided_node = rclcpp::Node::SharedPtr())
```

Comment:
I believe that when we do the copy assignment of the reference to the member variable at `node_ = provided_node;`, that is when the reference will get incremented. 

By passing in a reference, we only avoid adding a temporary reference in the function parameter itself.

In other words, by passing the shared ptr by value instead of by ref, we create the parameter `provided_node` and increment the reference count. Then assign it to the member variable and increment the reference count again. When the function returns the parameter is destroyed and the reference counted is decremented again. 

We'd end up taking the shared ptr lock an extra two times for the sole purpose of the temporary parameter object.